File: backend/app/api/routes/icons.py
from __future__ import annotations

import base64
from urllib.parse import urlparse

import httpx
from fastapi import APIRouter, Query
from fastapi.responses import Response

from app.utils.network import fetch_with_redirects


router = APIRouter(prefix="/icons", tags=["icons"])

# 1x1 transparent GIF
_TRANSPARENT_GIF = base64.b64decode(
    b"R0lGODlhAQABAPAAAP///wAAACH5BAAAAAAALAAAAAABAAEAAAICRAEAOw=="
)


def _build_headers(target_url: str) -> dict[str, str]:
    parsed = urlparse(target_url)
    origin = f"{parsed.scheme}://{parsed.netloc}" if parsed.scheme and parsed.netloc else ""
    return {
        "User-Agent": (
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
            "(KHTML, like Gecko) Chrome/119.0 Safari/537.36"
        ),
        "Accept": "image/avif,image/webp,image/apng,image/svg+xml,image/*;q=0.8,*/*;q=0.5",
        "Referer": origin or target_url,
    }


@router.get("/proxy")
async def proxy_icon(url: str = Query(..., description="Remote icon URL to proxy")) -> Response:
    """Proxy a remote icon and always return a successful image response.

    If the remote fetch fails or isn't an image, return a 1x1 transparent GIF
    with 200 to avoid noisy console errors in the frontend.
    """
    if not url:
        return Response(
            content=_TRANSPARENT_GIF,
            media_type="image/gif",
            headers={"Cache-Control": "public, max-age=86400"},
        )

    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            r = await fetch_with_redirects(client, url, headers=_build_headers(url))
            ctype = r.headers.get("content-type", "").lower()
            if r.status_code == 200 and ctype.startswith("image/") and r.content:
                return Response(
                    content=r.content,
                    media_type=ctype.split(";")[0] or "image/octet-stream",
                    headers={"Cache-Control": "public, max-age=86400"},
                )
    except Exception:
        # Swallow all errors and fall back to placeholder
        pass

    return Response(
        content=_TRANSPARENT_GIF,
        media_type="image/gif",
        headers={"Cache-Control": "public, max-age=86400"},
    )

File: backend-node/src/routes/icons.ts
/**
 * Icons API Routes (Favicon Fetching and Caching)
 */

import { FastifyInstance } from 'fastify';
import { JSDOM } from 'jsdom';
import { FeedRepository } from '../db/repositories/index.js';
import { getResponseHeader, requestBuffer, requestText } from '../services/outboundHttp.js';

const TRANSPARENT_GIF = Buffer.from(
  'R0lGODlhAQABAPAAAP///wAAACH5BAAAAAAALAAAAAABAAEAAAICRAEAOw==',
  'base64'
);

function buildProxyHeaders(targetUrl: string): Record<string, string> {
  try {
    const parsed = new URL(targetUrl);
    const origin = `${parsed.protocol}//${parsed.hostname}`;
    return {
      'User-Agent': 'Mozilla/5.0 (compatible; Aurora RSS Reader)',
      'Accept': 'image/avif,image/webp,image/apng,image/svg+xml,image/*;q=0.8,*/*;q=0.5',
      'Referer': origin || targetUrl,
    };
  } catch (error) {
    return {
      'User-Agent': 'Mozilla/5.0 (compatible; Aurora RSS Reader)',
      'Accept': 'image/avif,image/webp,image/apng,image/svg+xml,image/*;q=0.8,*/*;q=0.5',
      'Referer': targetUrl,
    };
  }
}

function resolveHref(href: string, baseUrl: URL): string {
  if (href.startsWith('http')) return href;
  if (href.startsWith('//')) return `${baseUrl.protocol}${href}`;
  if (href.startsWith('/')) return `${baseUrl.protocol}//${baseUrl.host}${href}`;
  return `${baseUrl.protocol}//${baseUrl.host}/${href}`;
}

async function discoverFaviconUrl(feedUrl: string): Promise<string> {
  const parsedFeedUrl = new URL(feedUrl);
  const domain = `${parsedFeedUrl.protocol}//${parsedFeedUrl.hostname}`;

  try {
    const faviconIcoUrl = `${domain}/favicon.ico`;
    const headResponse = await requestBuffer(faviconIcoUrl, {
      method: 'HEAD',
      maxRedirects: 3,
      maxRetries: 1,
      acceptedStatusCodes: [200],
      networkPolicy: 'public',
      maxResponseBytes: 0,
      headers: buildProxyHeaders(faviconIcoUrl),
      logContext: 'favicon-head',
    });

    const contentType = getResponseHeader(headResponse.headers, 'content-type');
    if (!contentType || contentType.startsWith('image/')) {
      return faviconIcoUrl;
    }
  } catch (error) {
    // Fall through to HTML parsing.
  }

  try {
    const response = await requestText(domain, {
      maxRedirects: 3,
      maxRetries: 1,
      acceptedStatusCodes: [200],
      networkPolicy: 'public',
      maxResponseBytes: 512 * 1024,
      headers: {
        'User-Agent': 'Mozilla/5.0 (compatible; Aurora RSS Reader)',
        'Accept': 'text/html,application/xhtml+xml',
      },
      logContext: 'favicon-html',
    });

    const dom = new JSDOM(response.body, { url: response.url });
    try {
      const selectors = [
        'link[rel="icon"]',
        'link[rel="shortcut icon"]',
        'link[rel="apple-touch-icon"]',
      ];

      for (const selector of selectors) {
        const href = dom.window.document.querySelector(selector)?.getAttribute('href');
        if (href) {
          return resolveHref(href, new URL(response.url));
        }
      }
    } finally {
      dom.window.close();
    }
  } catch (error) {
    // Fall through to Google fallback.
  }

  return `https://www.google.com/s2/favicons?domain=${parsedFeedUrl.hostname}&sz=64`;
}

export async function iconsRoutes(app: FastifyInstance) {
  const feedRepo = new FeedRepository();

  app.get('/icons/proxy', async (req, reply) => {
    const query = req.query as { url?: string };
    const targetUrl = query?.url;

    reply.header('Cache-Control', 'public, max-age=86400');

    if (!targetUrl) {
      reply.type('image/gif');
      return reply.send(TRANSPARENT_GIF);
    }

    try {
      const response = await requestBuffer(targetUrl, {
        headers: buildProxyHeaders(targetUrl),
        maxRedirects: 3,
        maxRetries: 1,
        maxResponseBytes: 512 * 1024,
        acceptedStatusCodes: [200],
        networkPolicy: 'public',
        logContext: 'icons-proxy',
      });

      const contentType = String(getResponseHeader(response.headers, 'content-type') || '').toLowerCase();
      if (contentType.startsWith('image/') && response.body.length > 0) {
        reply.type(contentType.split(';')[0]);
        return reply.send(response.body);
      }
    } catch (error) {
      // Fall through to transparent gif.
    }

    reply.type('image/gif');
    return reply.send(TRANSPARENT_GIF);
  });

  app.get('/icons/:feed_id', async (req, reply) => {
    const { feed_id } = req.params as { feed_id: string };
    const feed = feedRepo.findById(feed_id);

    if (!feed) {
      return { error: 'Feed not found' };
    }

    try {
      if (feed.favicon_url) {
        return {
          success: true,
          favicon_url: feed.favicon_url,
          cached: true,
        };
      }

      const faviconUrl = await discoverFaviconUrl(feed.url);
      feedRepo.update(feed_id, { favicon_url: faviconUrl });

      return {
        success: true,
        favicon_url: faviconUrl,
        cached: false,
      };
    } catch (error) {
      return {
        error: error instanceof Error ? error.message : 'Failed to fetch favicon',
      };
    }
  });

  app.delete('/icons/:feed_id', async (req) => {
    const { feed_id } = req.params as { feed_id: string };
    const feed = feedRepo.findById(feed_id);

    if (!feed) {
      return { error: 'Feed not found' };
    }

    feedRepo.update(feed_id, { favicon_url: null });
    return { success: true };
  });
}
